rtt_rosparam/include/rtt_rosparam/ros_param_data_source.hpp

```diff
@@ -0,0 +1,147 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the author nor the names of its contributors may
+ *    be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __RTT_ROSPARAM__ROS_PARAM_DATA_SOURCE_HPP
+#define __RTT_ROSPARAM__ROS_PARAM_DATA_SOURCE_HPP
+
+#include "rtt/internal/DataSource.hpp"
+#include "rtt/internal/DataSourceTypeInfo.hpp"
+#include "rtt/internal/Reference.hpp"
+
+#include "ros/ros.h"
+#include "ros/param.h"
+#include "ros/names.h"
+
+#include <boost/function.hpp>
+#include <exception>
+
+using namespace RTT;
+
+namespace rtt_rosparam {
+
+/**
+ * A DataSource which checks the ROS parameter server and it
+ * returns the parameter value in its get() method.
+ * It sets the parameter value in its set() method.
+ * The constructor receives the full qualified property name attached
+ * to the DataSource.
+ * <br><b>Warning</b>
+ * The properties with a RosParamDataSource<T> are not real-time safe
+ * and therefore this DataSource shouldn't be used in the updateHook()
+ * of a real-time component.
+ */
+template<typename T>
+class RosParamDataSource
+    : public internal::AssignableDataSource<T>
+{
+  /**
+   * Stores the full-qualified property name of the ROS parameter
+   */
+  std::string mparam_name_;
+  mutable typename internal::DataSource<T>::value_t mcached_data_;
+
+public:
+  /**
+   * Use shared_ptr.
+   */
+  typedef boost::intrusive_ptr< RosParamDataSource<T> > shared_ptr;
+
+  ~RosParamDataSource()
+  {
+  }
+
+  RosParamDataSource(std::string param_name_id)
+      : mparam_name_(param_name_id),
+        mcached_data_()
+  {
+  }
+
+  typename internal::DataSource<T>::result_t get() const {
+    (void) evaluate();
+    return value();
+  }
+
+  bool evaluate() const
+  {
+    T return_value;
+    if (!ros::param::getCached(mparam_name_, return_value)) {
```

Comment:
Avoid an unnecessary copy in line 98 below:
```suggestion
    if (!ros::param::getCached(mparam_name_, mcached_data_)) {
```